e2e/cypress/tests/integration/enterprise/self-hosted/self_hosted_purchase_spec.ts

```diff
@@ -0,0 +1,382 @@
+}
+function selectDropdownValue(placeholder: string, value: string) {
+    cy.contains(placeholder).click();
+    cy.contains(value).click();
+}
+
+function fillForm(form: PurchaseForm, currentUsers: Cypress.Chainable<number>) {
+    cy.uiGetPaymentCardInput().within(() => {
+        cy.get('[name="cardnumber"]').should('be.enabled').clear().type(form.card);
+        cy.get('[name="exp-date"]').should('be.enabled').clear().type(form.expires);
+        cy.get('[name="cvc"]').should('be.enabled').clear().type(form.cvc);
+    });
+
+    changeByPlaceholder('Organization Name', form.org);
+
+    changeByPlaceholder('Name on Card', form.name);
+    selectDropdownValue('Country', form.country);
+    changeByPlaceholder('Address', form.address);
+    changeByPlaceholder('City', form.city);
+    selectDropdownValue('State/Province', form.state);
+    changeByPlaceholder('Zip/Postal Code', form.zip);
+
+    if (form.agree) {
+        cy.get('#self_hosted_purchase_terms').click();
+    }
+
+    if (form === defaultSuccessForm) {
+        currentUsers.then((userCount) => {
+            cy.findByTestId('selfHostedPurchaseSeatsInput').clear().type((userCount + additionalSeatsToPurchase).toString());
+        });
+    } else if (form.seats) {
+        cy.findByTestId('selfHostedPurchaseSeatsInput').clear().type(form.seats.toString());
+    }
+
+    // while this will not work if the caller passes in an object
+    // that has member equality but not reference equality, this is
+    // good enough for the limited usage this function has
+    if (form === defaultSuccessForm) {
+        cy.contains('Upgrade').should('be.enabled');
+    }
+
+    return cy.contains('Upgrade');
+}
+
+function assertLine(lines: string[], key: string, value: string) {
+    const line = lines.find((line) => line.includes(key));
+    if (!line) {
+        throw new Error('Expected license to show start date line but did not');
+    }
+    if (!line.includes(value)) {
+        throw new Error(`Expected license ${key} of ${value}, but got ${line}`);
+    }
+}
+
+function getCurrentUsers(): Cypress.Chainable<number> {
+    return cy.request('/api/v4/analytics/old?name=standard&team_id=').then((response) => {
+        const userCount = response.body.find((row: Cypress.AnalyticsRow) => row.name === 'unique_user_count');
+        return userCount.value;
+    });
+}
+
+describe('Self hosted Purchase', () => {
+    let adminUser: Cypress.UserProfile | undefined;
+
+    before(() => {
+        cy.apiInitSetup().then(() => {
+            cy.apiAdminLogin().then((result) => {
+                // assertion because current typings are wrong.
+                adminUser = (result as unknown as {user: Cypress.UserProfile}).user;
+                cy.apiDeleteLicense();
+                cy.visit('/');
+            });
+        });
+    });
+
+    it('happy path, can purchase a license and have it applied automatically', () => {
+        cy.apiAdminLogin();
+        cy.apiDeleteLicense();
```

Comment:
@neallred do you know if there's an api that we can block/alter to make it seem as 'instance' without license.
If we eventually, run this along with other tests, it permanently remove license in CI

Reply:
I think we could fake it for this test, but that we wouldn't want to because even if we did, this test still alters the license from what other tests expect (professional plan is purchased and applied, but other tests assume an enterprise level license). Not sure how doable this is with the existing test framework, but I think a more robust approach would be that after this test runs, it restores the license that all other tests depend on. Running this test automatically is also blocked by  https://mattermost.atlassian.net/browse/MM-49161 .

Reply:
Agree that that is ideal, but the current test run setup makes assumptions on that license and I think lacks the setup to make sure the right license is applied. So I think it makes sense to sort this to the end of the run group: https://git.internal.mattermost.com/qa/cypress-ui-automation/-/blob/e2e-master-prod/.gitlab-ci.yml#L44.